be/src/runtime/load_channel.cpp

```diff
@@ -0,0 +1,163 @@
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "runtime/load_channel.h"
+
+#include "runtime/mem_tracker.h"
+#include "runtime/tablets_channel.h"
+#include "olap/lru_cache.h"
+
+namespace doris {
+
+LoadChannel::LoadChannel(const UniqueId& load_id, int64_t mem_limit)
+        : _load_id(load_id) {
+    _lastest_success_channel = new_lru_cache(8);
+    _mem_tracker.reset(new MemTracker(mem_limit, _load_id.to_string()));
+    // _last_updated_time should be set before being inserted to
+    // _load_channels in load_channel_mgr, or it may be erased
+    // immediately by gc thread.
+    _last_updated_time = time(nullptr);
+}
+
+LoadChannel::~LoadChannel() {
+    LOG(INFO) << "load channel mem peak usage: " << _mem_tracker->peak_consumption()
+        << ", info: " << _mem_tracker->debug_string()
+        << ", load id: " << _load_id;
+    delete _lastest_success_channel;
+
+    std::lock_guard<std::mutex> l(_lock);
+}
+
+Status LoadChannel::open(const PTabletWriterOpenRequest& params) {
+    int64_t index_id = params.index_id();
+    std::shared_ptr<TabletsChannel> channel;
+    {
+        std::lock_guard<std::mutex> l(_lock);
+        auto it = _tablets_channels.find(index_id);
+        if (it != _tablets_channels.end()) {
+            channel = it->second;
+        } else {
+            // create a new tablets channel
+            TabletsChannelKey key(params.id(), index_id);
+            channel.reset(new TabletsChannel(key, _mem_tracker.get()));
+            _tablets_channels.insert({index_id, channel});
+        }
+    }
+
+    RETURN_IF_ERROR(channel->open(params));
+
+    if (!_opened) {
+        _opened = true;
+    }
+    _last_updated_time = time(nullptr);
+    return Status::OK();
+}
+
+static void dummy_deleter(const CacheKey& key, void* value) {
+}
+
+Status LoadChannel::add_batch(
+        const PTabletWriterAddBatchRequest& request,
+        google::protobuf::RepeatedPtrField<PTabletInfo>* tablet_vec) {
+
+    int64_t index_id = request.index_id();
+    // 1. get tablets channel
+    std::shared_ptr<TabletsChannel> channel;
+    {
+        std::lock_guard<std::mutex> l(_lock);
+        auto it = _tablets_channels.find(index_id);
+        if (it == _tablets_channels.end()) {
+            auto handle = _lastest_success_channel->lookup(std::to_string(index_id));
+            // success only when eos be true
+            if (handle != nullptr && request.has_eos() && request.eos()) {
```

Comment:
If use _finished_channels, no need to check if request is eos

Reply:
OK